### PR TITLE
Added lazy optional handlers, closes #7

### LINKED ISSUE #7: [1.14] LazyOptionals shouldn't be created every call
As seen here: https://github.com/InnovativeOnlineIndustries/Titanium/blob/1.14/src/main/java/com/hrznstudio/titanium/block/tile/TileActive.java#L122-L137 y'all are creating the optionals every time, shouldn't do this.

### PR DESCRIPTION
**Pull request by Buuz135**

### CODE REVIEW

File: src/main/java/com/hrznstudio/titanium/block/tile/sideness/ICapabilityHolder.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import net.minecraftforge.common.util.LazyOptional;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+
+public interface ICapabilityHolder<T extends Object, R extends Object> {
+
+    void add(T capability);
+
+    HashMap<FacingUtil.Sideness, LazyOptional<R>> getCapabilities();
```
Comment (SkySom) [inline src/main/java/com/hrznstudio/titanium/block/tile/sideness/ICapabilityHolder.java]: Forcing an implementation in the interface. Bad.

File: src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java [hunk 2]
```diff
@@ -47,7 +61,29 @@ public MultiInvCapabilityHandler getCapabilityForSide(FacingUtil.Sideness sidene
...
+    }
+
+    @Override
+    public HashMap<FacingUtil.Sideness, LazyOptional<MultiInvCapabilityHandler>> getCapabilities() {
+        return lazyOptionals;
+    }
+
+    @Override
+    public LazyOptional<MultiInvCapabilityHandler> getCapabilityForSide(FacingUtil.Sideness sideness) {
+        return lazyOptionals.get(sideness);
```
Comment (SkySom) [inline src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java:71]: You should make this as Nonnull and ensure it is such.

File: src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java [hunk 3]
```diff
@@ -10,33 +10,47 @@
...
 
-    public void addInventory(PosInvHandler invHandler) {
+    @Override
+    public void add(PosInvHandler invHandler) {
         this.inventoryHandlers.add(invHandler);
+        rebuildCapability(new FacingUtil.Sideness[]{null});
+        rebuildCapability(FacingUtil.Sideness.values());
+    }
+
+    private void rebuildCapability(FacingUtil.Sideness... sides){
```
Comment (SkySom) [inline src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java]: Why make this varargs if you always create the arrays.

File: src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java [hunk 4]
```diff
@@ -10,33 +10,47 @@
...
+        for (FacingUtil.Sideness side : sides) {
+            lazyOptionals.put(side, getCapabilityForSide(side).map(multiInvCapabilityHandler -> new MultiInvCapabilityHandler(getHandlersForSide(side))));
+        }
     }
 
-    public MultiInvCapabilityHandler getCapabilityForSide(FacingUtil.Sideness sideness) {
+    private List<PosInvHandler> getHandlersForSide(FacingUtil.Sideness sideness){
         if (sideness == null)
-            return new MultiInvCapabilityHandler(new ArrayList<>(inventoryHandlers));
+            return new ArrayList<>(inventoryHandlers);
```
Comment (SkySom) [inline src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java:53]: So are brackets required or not required for single line ifs. There is no consistency in which to use. Same line or not?

File: src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java [hunk 5]
```diff
@@ -47,7 +61,29 @@ public MultiInvCapabilityHandler getCapabilityForSide(FacingUtil.Sideness sidene
...
+
+    @Override
+    public LazyOptional<MultiInvCapabilityHandler> getCapabilityForSide(FacingUtil.Sideness sideness) {
+        return lazyOptionals.get(sideness);
+    }
+
+    @Override
+    public boolean handleFacingChange(String handlerName, FacingUtil.Sideness facing, IFacingHandler.FaceMode mode) {
+        for (PosInvHandler inventoryHandler : inventoryHandlers) {
+            if (inventoryHandler.getName().equals(handlerName) && inventoryHandler instanceof IFacingHandler){
```
Comment (SkySom) [inline src/main/java/com/hrznstudio/titanium/block/tile/inventory/MultiInventoryHandler.java]: Space before the bracket or not? Watch your formatting.

File: src/main/java/com/hrznstudio/titanium/block/tile/TileActive.java [hunk 6]
```diff
@@ -66,7 +68,7 @@ public TileActive(BlockTileBase base) {
     @Override
     public boolean onActivated(PlayerEntity playerIn, Hand hand, Direction facing, double hitX, double hitY, double hitZ) {
         if (multiTankHandler != null) {
-            return FluidUtil.interactWithFluidHandler(playerIn, hand, multiTankHandler.getCapabilityForSide(null));
+            return FluidUtil.interactWithFluidHandler(playerIn, hand, multiTankHandler.getCapabilityForSide(null).orElse(new MultiTankHandler.MultiTankCapabilityHandler(new ArrayList<>())));
```
Comment (SkySom) [inline src/main/java/com/hrznstudio/titanium/block/tile/TileActive.java:71]: What is the point of creating a new Handler here?

File: src/main/java/com/hrznstudio/titanium/block/tile/fluid/MultiTankHandler.java [hunk 7]
```diff
@@ -10,59 +10,95 @@
...
+        return handlers;
+    }
 
+    @Override
+    public HashMap<FacingUtil.Sideness, LazyOptional<MultiTankCapabilityHandler>> getCapabilities() {
+        return lazyOptionals;
+    }
+
+    @Override
+    public LazyOptional<MultiTankCapabilityHandler> getCapabilityForSide(@Nullable FacingUtil.Sideness sideness) {
```
Comment (SkySom) [inline src/main/java/com/hrznstudio/titanium/block/tile/fluid/MultiTankHandler.java:70]: Mark nonnull and ensure such.

**Review by SkySom — CHANGES_REQUESTED** [review]

File: src/main/java/com/hrznstudio/titanium/block/tile/TileActive.java (hunk 6 above)
Comment (Buuz135) [reply src/main/java/com/hrznstudio/titanium/block/tile/TileActive.java:71]: The capability can't be null in that method

**Review by SkySom — APPROVED** [review]

**Review by adrwa — APPROVED** [review]

**Pull request merged**